### PR TITLE
Upgrade to Gitlab 9.3.0

### PR DESCRIPTION
**Pull request by timfeirg**
Should be close to a release, thought I'd create a PR first.

### CODE REVIEW

**Comment by Ilya-Kuchaev** [comment]
9.2.7 is already merged
https://github.com/sameersbn/docker-gitlab/pull/1258

**Comment by xen0n** [comment]
Note: This release [contains Aliyun OSS support for backups](https://gitlab.com/gitlab-org/gitlab-ce/blob/v9.3.0/CHANGELOG.md). According to [the corresponding MR](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/11849) it should be a matter of generating the respective fog configuration. It'd be better to include the feature as well!

**Comment by timfeirg** [comment]
I'm unable to look into fog-aliyun and decide which of the configurations should be added into docker-gitlab. Those who actually use this will do a better job. @xen0n

**Comment by timfeirg** [comment]
travis build has failed due to long time no output, so I'll close and re-open to trigger a rebuild.

**Comment by timfeirg** [comment]
travis failed again, at https://github.com/sameersbn/docker-gitlab/blob/master/assets/build/install.sh#L137
So I've removed the output redirection and let it talk, and see if travis is good with that.

**Comment by reynold-lariza** [comment]
Hey guys, 

Apart from that Aliyun OSS support, how goes your testing with 9.3.0? any issues found? I just upgraded to 9.2.7 yesterday. 

I get more cautious as our usage of Gitlab CE grows ;) I wouldn't want to bang my head on the wall if I break it :P

**Comment by timfeirg** [comment]
Not extremely sure about testing here.

The purpose of docker-gitlab is to package gitlab and all its components for ease of deployment, not testing gitlab or fixing its bug.

So once the setup and the migration is working, docker-gitlab itself should not require any more testing, at least for minor releases which doesn't Chang the way gitlab is deployed.

And so I didn't test this PR. @reynold-lariza
Cc @solidnerd

**Comment by reynold-lariza** [comment]
Hey @timfeirg,

I mean you have upgraded your own instance of gitlab to 9.3.0, you are able to upgrade successfully without any smoke or noise at all? You made this PR with that kind of confidence right? ;)

**Comment by timfeirg** [comment]
I haven't, as explained in https://github.com/sameersbn/docker-gitlab/pull/1260#issuecomment-310956633

**Comment by hugomd** [comment]
Is this ready to be merged?

**Comment by timfeirg** [comment]
ping @solidnerd

**Comment by xen0n** [comment]
I think I'll do the Aliyun part myself after this get merged. Considering the release frequency of GitLab the resulting minor gap of support seems negligible.

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 FROM sameersbn/ubuntu:14.04.20170503
 MAINTAINER sameer@damagehead.com
 
-ENV GITLAB_VERSION=9.2.7 \
+ENV GITLAB_VERSION=9.3.0 \
     RUBY_VERSION=2.3 \
     GOLANG_VERSION=1.6.3 \
```
Comment (jostyee) [inline Dockerfile]:
```
    GITLAB_SHELL_VERSION=5.0.5 \
    GITLAB_WORKHORSE_VERSION=2.1.1 \
    GITLAB_PAGES_VERSION=0.4.3 \
```

https://gitlab.com/gitlab-org/gitlab-ce/tree/acb7f257db65e8ce2907e4de7c42627c110476e1

Comment (timfeirg) [inline Dockerfile]: PR updated, thanks.

**Comment by solidnerd** [comment]
Hey, 
what is currently missing for me the changesets for all the [config files](https://github.com/sameersbn/docker-gitlab/tree/master/assets/runtime/config) . Could you check if there are some changes or there are features to adept through configuration via environment variables.

**Comment by solidnerd** [comment]
ping @timfeirg

**Comment by timfeirg** [comment]
Working on this, I'm pulling out the diff from gitlab ce between 9.2.7 and 9.3.0 to check if any configuration needs changing.

Wish there's some sort of tutorial or contribution guideline to include this sort of knowledge, I mean, things to watch out for when making a PR to upgrade docker-gitlab, this would definitely help inexperienced contributors like myself.

**Comment by solidnerd** [comment]
I will work on this ;)

**Comment by timfeirg** [comment]
I have:

* compared [gitlab-shell config](https://github.com/sameersbn/docker-gitlab/blob/master/assets/runtime/config/gitlab-shell/config.yml) to [gitlab-shell config example](https://gitlab.com/gitlab-org/gitlab-shell/blob/master/config.yml.example)
* compared [gitab ce config](https://github.com/sameersbn/docker-gitlab/tree/master/assets/runtime/config/gitlabhq) to [gitlab ce config examples](https://gitlab.com/gitlab-org/gitlab-ce/tree/master/config)

As for nginx configurations, I'm not sure how to conclude the changes from the massive diff. Got pretty much nothing from the changelog. How would you approach this? @solidnerd

**Comment by smuething** [comment]
There is now an additional step during the installation of gitlab-shell: You need to run `bin/compile` before `bin/install` to compile two Go binaries. For that to work, the Go environment has to be available to `bin/compile`. Without this extra step, SSH access to repositories will fail after the update.

**Comment by smuething** [comment]
Sorry, forgot to mention: As of GitLab 9.3.0, upstream now officially requires Go >= 1.8.3

**Comment by Ilya-Kuchaev** [comment]
Meanwhile https://gitlab.com/gitlab-org/gitlab-ce/blob/master/CHANGELOG.md#932-2017-06-27 is out :(

File: assets/build/install.sh [hunk 2]
```diff
@@ -55,6 +55,12 @@ GITLAB_SHELL_VERSION=${GITLAB_SHELL_VERSION:-$(cat ${GITLAB_INSTALL_DIR}/GITLAB_
 GITLAB_WORKHORSE_VERSION=${GITLAB_WORKHOUSE_VERSION:-$(cat ${GITLAB_INSTALL_DIR}/GITLAB_WORKHORSE_VERSION)}
 GITLAB_PAGES_VERSION=${GITLAB_PAGES_VERSION:-$(cat ${GITLAB_INSTALL_DIR}/GITLAB_PAGES_VERSION)}
 
+#download golang
+echo "Downloading Go ${GOLANG_VERSION}..."
+wget -cnv https://storage.googleapis.com/golang/go${GOLANG_VERSION}.linux-amd64.tar.gz -P ${GITLAB_BUILD_DIR}/
+tar -xf ${GITLAB_BUILD_DIR}/go${GOLANG_VERSION}.linux-amd64.tar.gz -C /tmp/
+ln -s /tmp/go/bin/go /usr/local/bin/go
```
Comment (timfeirg) [inline assets/build/install.sh]: using `bin/compile` poses a problem, they use ruby script to run `go install gitlab.com/gitlab-org/gitlab-shell/go/cmd/...`, and when the script does this, it didn't inherit `$PATH` and `$GOROOT` from docker-gitlab's install.sh, and the go executables cannot be found.

So I had to make go available in the default `$PATH`.

@solidnerd

Comment (phpb-com) [inline assets/build/install.sh]: Try something like this:

```diff
+if [[ -x ./bin/compile ]]; then
+  echo "Compiling gitlab-shell golang executables..."
+  exec_as_git PATH=/tmp/go/bin:$PATH GOROOT=/tmp/go ./bin/compile
+fi
```

Comment (timfeirg) [inline assets/build/install.sh]: this blows my mind @phpb-com

**Comment by phpb-com** [comment]
You'll also need to install gitaly, you can check my fork to see how you can do it, if you want.

**Comment by timfeirg** [comment]
Absolutely, but I didn't see a docker-gitlab fork on your github @phpb-com

**Comment by abeltay** [comment]
He uses this repo https://gotfix.com/docker/gitlab/

**Comment by phpb-com** [comment]
The mirror of my fork on github is https://github.com/gotfix/gitlab @timfeirg

**Comment by jostyee** [comment]
LGTM. I suggest remove the WIP tag so @solidnerd can review it.

**Comment by timfeirg** [comment]
I'm still trying to test this, just having a hard time setting things up this using censored internet.

**Comment by solidnerd** [comment]
@timfeirg Hey, thanks for your awesome work. I will gonna review it tomorrow on Sunday morning. I assume that we can release it tomorrow. More Changes from other PR's will follow as well.

**Comment by timfeirg** [comment]
I've tested my commit and verified that ssh clone is working, along with the gitlab web server. I'm not sure if that actually means gitaly and gitlab-shell is working correctly.

File: assets/build/install.sh [hunk 3]
```diff
@@ -134,7 +153,7 @@ exec_as_git yarn install --production --pure-lockfile
 
 echo "Compiling assets. Please be patient, this could take a while..."
 #Adding webpack compile needed since 8.17
-exec_as_git bundle exec rake assets:clean assets:precompile webpack:compile USE_DB=false SKIP_STORAGE_VALIDATION=true >/dev/null 2>&1
+exec_as_git bundle exec rake assets:clean assets:precompile webpack:compile USE_DB=false SKIP_STORAGE_VALIDATION=true
```
Comment (solidnerd) [inline assets/build/install.sh]: Could make these changes as a separate commit. I assume that we should change the precompile method as well to
```bash
exec_as_git bundle exec rake gitlab:assets:compile USE_DB=false SKIP_STORAGE_VALIDATION=true
```

File: Dockerfile [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
 MAINTAINER sameer@damagehead.com
 
-ENV GITLAB_VERSION=9.2.7 \
+ENV GITLAB_VERSION=9.3.0 \
     RUBY_VERSION=2.3 \
-    GOLANG_VERSION=1.6.3 \
-    GITLAB_SHELL_VERSION=5.0.4 \
-    GITLAB_WORKHORSE_VERSION=2.0.0 \
-    GITLAB_PAGES_VERSION=0.4.2 \
+    GOLANG_VERSION=1.8.3 \
```
Comment (solidnerd) [inline Dockerfile]: For all the GitLab Components like GITLAB_SHELL_VERSION could you make a separate commit for the version changes. And more important for the GOLANG_VERSION as well.

File: assets/build/install.sh (hunk 3 above)
Comment (timfeirg) [inline assets/build/install.sh]: I'm not sure I understand. Do you mean this?
```
diff --git a/assets/runtime/functions b/assets/runtime/functions
index 7d4a647..d372861 100644
--- a/assets/runtime/functions
+++ b/assets/runtime/functions
@@ -1525,7 +1525,7 @@ migrate_database() {
     # assets need to be recompiled when GITLAB_RELATIVE_URL_ROOT is used
     if [[ -n ${GITLAB_RELATIVE_URL_ROOT} ]]; then
       echo "Recompiling assets (relative_url in use), this could take a while..."
-      exec_as_git bundle exec rake assets:clean assets:precompile webpack:compile  >/dev/null 2>&1
+      exec_as_git bundle exec rake gitlab:assets:compile USE_DB=false SKIP_STORAGE_VALIDATION=true
     fi

     echo "Clearing cache..."
```
Meaning that `gitlab:assets:compile` equals `assets:precompile webpack:compile`?
@solidnerd

Comment (solidnerd) [inline assets/build/install.sh]: It seems so. In the upstream documentation they are using only `gitlab:assets:compile` to compile the assets.

File: Dockerfile (hunk 4 above)
Comment (jostyee) [inline Dockerfile]: @solidnerd Separate ci or separate pr? but what for? These version updates are required by GitLab CE.

Comment (solidnerd) [inline Dockerfile]: Yeah. I want only fine grained commits for theses changes. Then we get a simple history that everybody can understand directly and doesn't need to check every upgrade commit. I will make a separate commit for every release. So from this point we could see which changes happen between a release.
